src/rdo.rs

```diff
@@ -1100,67 +1103,216 @@ pub fn rdo_partition_decision(
+                cost_acc += cost[pli];
+              }
+              RestorationFilter::Wiener{..} => unreachable!() // coming soon
+            }
+          }
+          if best_cost_acc < 0. || cost_acc < best_cost_acc {
+            cdef_change = true;
+            best_cost_acc = cost_acc;
+            best_index = cdef_index;
+            for pli in 0..3 {
+              best_cost[pli] = cost[pli];
             }
           }
         }
       }
     }
 
-    if cdef_index == 0 || err < best_err {
-      best_err = err;
-      best_index = cdef_index;
+    if !cdef_change && !first_loop { break; }
+    first_loop = false;
+    
+    // check for new best restoration filter if enabled
+    let mut lrf_change = false;
+    if fi.sequence.enable_restoration && lrf_any_uncoded {
+      // need cdef output from best index, not just last iteration
+      if fi.sequence.enable_cdef {
+        cdef_filter_superblock(fi, &mut cdef_input, &mut lrf_input,
+                               &mut cw.bc, &sbo_0, &sbo, best_index as u8, &cdef_dirs);
+      }
+
+      // Wiener LRF decision coming soon
+
+      // SgrProj LRF decision
+      for pli in 0..3 {
+        let ru = fs.restoration.plane[pli].restoration_unit(sbo);
+        if !ru.coded {
+          for set in 0..16 {
+            let in_plane = &fs.input.planes[pli];  // reference
+            let ipo = &sbo.plane_offset(&in_plane.cfg);
+            let cdef_plane = &lrf_input.planes[pli];
+            let (xqd0, xqd1) = sgrproj_solve(set, fi,
+                                             &in_plane.slice(ipo),
+                                             &cdef_plane.slice(&PlaneOffset{x: 0, y: 0}),
+                                             cmp::min(cdef_plane.cfg.width, fi.width - ipo.x as usize),
+                                             cmp::min(cdef_plane.cfg.height, fi.height - ipo.y as usize));
+            let current_lrf = RestorationFilter::Sgrproj{set: set, xqd: [xqd0, xqd1]};
+            if let RestorationFilter::Sgrproj{set, xqd} = current_lrf {
+              // At present, this is a gross simplification
+              sgrproj_stripe_filter(set, xqd, fi,
+                                    lrf_input.planes[pli].cfg.width, lrf_input.planes[pli].cfg.height,
+                                    lrf_input.planes[pli].cfg.width, lrf_input.planes[pli].cfg.height,
+                                    &lrf_input.planes[pli].slice(&PlaneOffset{x:0, y:0}),
+                                    &lrf_input.planes[pli].slice(&PlaneOffset{x:0, y:0}),
+                                    &mut lrf_output.planes[pli].mut_slice(&PlaneOffset{x:0, y:0}));
+            }
+            let err = rdo_loop_plane_error(sbo, fi, fs, &cw.bc, &lrf_output, pli);
+            let rate = cw.count_lrf_switchable(w, &fs.restoration, &current_lrf, pli);
+            let cost = err as f64 + get_lambda(fi) * rate as f64 / ((1<<OD_BITRES) as f64);
```

Comment:
This line needs to be updated now that rate-control has landed.
```suggestion
            let cost = err as f64 + fi.lambda * rate as f64 / ((1<<OD_BITRES) as f64);
```